sw/device/tests/csrng_edn_cmd_test.c

```diff
@@ -0,0 +1,243 @@
+
+  CHECK_DIF_OK(dif_rv_plic_irq_set_enabled(
+      &plic, irq_id, kTopEarlgreyPlicTargetIbex0, kDifToggleEnabled));
+  CHECK_DIF_OK(dif_rv_plic_target_set_threshold(
+      &plic, kTopEarlgreyPlicTargetIbex0, /*threshold=*/0u));
+
+  CHECK_DIF_OK(dif_csrng_irq_set_enabled(&csrng, kDifCsrngIrqCsCmdReqDone,
+                                         kDifToggleEnabled));
+
+  irq_global_ctrl(true);
+  irq_external_ctrl(true);
+}
+
+/**
+ * Blocks until a CSRNG entropy request interrupt is received.
+ */
+static void csrng_cmd_done_irq_block_wait(void) {
+  // The interrupt can come before we enter sleep, so we check beforehand.
+  while (true) {
+    if (csrng_cmd_done_isr_set) {
+      break;
+    }
+    wait_for_interrupt();
+  }
+  CHECK_DIF_OK(dif_csrng_irq_set_enabled(&csrng, kDifCsrngIrqCsCmdReqDone,
+                                         kDifToggleDisabled));
+}
+
+/**
+ * Requests data from CSRNG software instance.
+ *
+ * Asserts error if there are any repeated words in the output data, or if there
+ * are any errors set in the CSRNG status registers.
+ */
+static void csrng_generate_output_check(void) {
+  uint32_t output[kTestParamFifoBufferSize] = {0};
+  csrng_testutils_cmd_generate_run(&csrng, output, ARRAYSIZE(output));
+
+  uint32_t prev_data = 0;
+  for (size_t i = 0; i < ARRAYSIZE(output); ++i) {
+    CHECK(prev_data != output[i],
+          "Received duplicate data. Last index: %d, value: 0x%x", i, prev_data);
+  }
+}
+
+static dif_edn_seed_material_t edn_seed_material_build(void) {
+  dif_edn_seed_material_t seed;
+  seed.len =
+      rand_testutils_gen32_range(/*min=*/0, kDifEntropySeedMaterialMaxWordLen);
+  for (size_t i = 0; i < seed.len; ++i) {
+    seed.data[i] = rand_testutils_gen32();
+  }
+  return seed;
+}
+
+static dif_edn_auto_params_t edn_auto_params_build(void) {
+  dif_edn_seed_material_t seed0 = edn_seed_material_build();
+  dif_edn_seed_material_t seed1 = edn_seed_material_build();
+  return (dif_edn_auto_params_t){
+      .instantiate_cmd =
+          {
+              .cmd = csrng_cmd_header_build(kCsrngAppCmdInstantiate,
+                                            kDifCsrngEntropySrcToggleEnable,
+                                            seed0.len,
+                                            /*generate_len=*/0),
+              .seed_material = seed0,
+          },
+      .reseed_cmd =
+          {
+              .cmd = csrng_cmd_header_build(kCsrngAppCmdReseed,
+                                            kDifCsrngEntropySrcToggleEnable,
+                                            seed1.len,
+                                            /*generate_len=*/0),
+              .seed_material = seed1,
+          },
+      .generate_cmd =
+          {
+              .cmd = csrng_cmd_header_build(kCsrngAppCmdGenerate,
+                                            kDifCsrngEntropySrcToggleEnable,
+                                            /*cmd_len=*/0,
```

Comment:
ah i was about to say, this length parameter doesn't seem quite right, but looks like it's been fixed already!

Reply:
Fixed :)